pkg/kubelet/util/pod_startup_latency_tracker.go

```diff
@@ -98,27 +102,85 @@ func (p *basicPodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when tim
+			excludedTimeStart = state.firstStartedPulling
+		} else if !state.firstInitContainerStart.IsZero() {
+			excludedTimeStart = state.firstInitContainerStart
+		}
+
+		if !excludedTimeStart.IsZero() {
+			preExcludedDuration := excludedTimeStart.Sub(pod.CreationTimestamp.Time)
+			if preExcludedDuration > 0 {
+				podStartSLOduration += preExcludedDuration
+			}
+		}
+
+		// Add gap between image pulling end and init container start if there is any
+		if !state.lastFinishedPulling.IsZero() && !state.firstInitContainerStart.IsZero() {
+			// Only count gap if init container starts after image pulling ends (no overlap)
+			if state.firstInitContainerStart.After(state.lastFinishedPulling) {
+				gapDuration := state.firstInitContainerStart.Sub(state.lastFinishedPulling)
+				if gapDuration > 0 {
+					podStartSLOduration += gapDuration
+				}
+			}
+		}
+
+		// Add time from last dependency completion to containers running
+		if state.lastFinishedPulling.After(state.lastInitContainerFinish) {
+			excludedTimeEnd = state.lastFinishedPulling
+		} else if !state.lastInitContainerFinish.IsZero() {
+			excludedTimeEnd = state.lastInitContainerFinish
+		} else if !state.lastFinishedPulling.IsZero() {
+			excludedTimeEnd = state.lastFinishedPulling
+		}
+
+		if !excludedTimeEnd.IsZero() {
+			postExcludedDuration := when.Sub(excludedTimeEnd)
+			if postExcludedDuration > 0 {
+				podStartSLOduration += postExcludedDuration
+			}
+		} else if excludedTimeStart.IsZero() {
+			// No dependencies at all, count entire duration
+			podStartSLOduration = podStartingDuration
+		}
+
+		isStatefulPod := isStatefulPod(pod)
 
 		klog.InfoS("Observed pod startup duration",
 			"pod", klog.KObj(pod),
-			"podStartSLOduration", podStartSLOduration,
+			"podStartSLOduration", podStartSLOduration.Seconds(),
 			"podStartE2EDuration", podStartingDuration,
+			"isStatefulPod", isStatefulPod,
 			"podCreationTimestamp", pod.CreationTimestamp.Time,
 			"firstStartedPulling", state.firstStartedPulling,
 			"lastFinishedPulling", state.lastFinishedPulling,
+			"firstInitContainerStart", state.firstInitContainerStart,
+			"lastInitContainerFinish", state.lastInitContainerFinish,
 			"observedRunningTime", state.observedRunningTime,
 			"watchObservedRunningTime", when)
 
-		metrics.PodStartSLIDuration.WithLabelValues().Observe(podStartSLOduration)
-		metrics.PodStartTotalDuration.WithLabelValues().Observe(podStartingDuration.Seconds())
-		state.metricRecorded = true
-		// if is the first Pod with network track the start values
-		// these metrics will help to identify problems with the CNI plugin
-		if !pod.Spec.HostNetwork && !p.firstNetworkPodSeen {
-			metrics.FirstNetworkPodStartSLIDuration.Set(podStartSLOduration)
-			p.firstNetworkPodSeen = true
+		if !isStatefulPod {
+			metrics.PodStartSLIDuration.WithLabelValues().Observe(podStartSLOduration.Seconds())
+			metrics.PodStartTotalDuration.WithLabelValues().Observe(podStartingDuration.Seconds())
```

Comment:
I don't think it is a good idea to only record this metric for stateless pod, unless there is another metric to record their latency, we won't record them anymore with this change.
I think it is good to keep PodStartTotalDuration as a catch all latency metric that records the end-to-end pod startup latency. It is a good metric to catch issues happening outside of kubelet's control.